xcpmd/src/screen-module.c

```diff
@@ -0,0 +1,320 @@
+/*
+ * screen-module.c
+ *
+ * XCPMD module that provides display power management actions.
+ *
+ * Copyright (c) 2015 Assured Information Security, Inc.
+ *
+ * Author:
+ * Jennifer Temkin <temkinj@ainfosec.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include "project.h"
+#include "xcpmd.h"
+#include "rpcgen/surfman_client.h"
+#include "rules.h"
+#include <math.h>
+
+//Function prototypes
+void screen_on(struct arg_node * args);
+void screen_off(struct arg_node * args);
+void set_backlight(struct arg_node * args);
+void increase_backlight(struct arg_node * args);
+void decrease_backlight(struct arg_node * args);
+static int get_brightness(void);
+static int get_raw_brightness(int *brightness, int *max_brightness);
+static void set_brightness(int percentage);
+
+
+//Private data structures
+struct action_table_row {
+    char * name;
+    void (* func)(struct arg_node *);
+    char * prototype;
+    char * pretty_prototype;
+};
+
+
+//Private data
+static struct action_table_row action_table[] = {
+    {"screenOn"          , screen_on          , "n" , "void"                    } ,
+    {"screenOff"         , screen_off         , "n" , "void"                    } ,
+    {"setBacklight"      , set_backlight      , "i" , "int backlight_percent"   } ,
+    {"increaseBacklight" , increase_backlight , "i" , "int percent_to_increase" } ,
+    {"decreaseBacklight" , decrease_backlight , "i" , "int percent_to_decrease" }
+};
+
+static unsigned int num_action_types = sizeof(action_table) / sizeof(action_table[0]);
+static int times_loaded = 0;
+
+
+//Registers this module's action types.
+//The constructor attribute causes this function to run at load (dlopen()) time.
+__attribute__ ((constructor)) static void init_module() {
+
+    unsigned int i;
+
+    if (times_loaded > 0)
```

Comment:
I also thought that, but dlopen() has some very counterintuitive behavior with shadowing symbols. Declaring the constructor as static should prevent it from being an issue. I'll remove the remnants.